lib/domain/services/execution_services.dart

```diff
@@ -0,0 +1,163 @@
+      if (pristineMemos.isNotEmpty) {
+        executableMemosChunk.add(pristineMemos.first);
+        pristineMemos.removeAt(0);
+      } else {
+        executableMemosChunk.add(executedMemoPerStability[sortedStabilityKeys]!);
+        executedMemoPerStability.remove(sortedStabilityKeys);
+      }
+    }
+
+    return executableMemosChunk;
+  }
+
+  @override
+  Future<void> addExecutions(List<MemoExecution> executions, {required String collectionId}) async {
+    final associatedData = await Future.wait([
+      memoRepo.getAllMemos(collectionId: collectionId),
+      userRepo.getUser(),
+      collectionRepo.getCollection(id: collectionId),
+    ]);
+
+    var totalTimeSpent = 0;
+    var newUniqueMemos = 0;
+    final executionsAmounts = <MemoDifficulty, int>{};
+    final updatedMemos = <Memo>[];
+
+    final allCollectionMemos = associatedData[0] as List<Memo>;
+    executions.forEach((execution) {
+      totalTimeSpent += execution.timeSpentInMillis;
+
+      // Creates a copy of the memo associated with this execution (same id) and with the new execution-related
+      // properties
+      final associatedMemo = allCollectionMemos.firstWhere((memo) => memo.id == execution.memoId);
+      final associatedMemoExecutionAmounts = associatedMemo.executionsAmounts;
+      final memoCurrentValue = associatedMemoExecutionAmounts[execution.markedDifficulty] ?? 0;
+      associatedMemoExecutionAmounts[execution.markedDifficulty] = memoCurrentValue + 1;
+      final updatedMemo = associatedMemo.copyWith(
+        execution,
+        associatedMemoExecutionAmounts,
+        associatedMemo.timeSpentInMillis + execution.timeSpentInMillis,
+      );
+      updatedMemos.add(updatedMemo);
+
+      // Update the unique count if this memo has never been executed before
+      if (associatedMemo.isPristine) {
+        newUniqueMemos++;
+      }
+
+      // Update the total count for the specific difficulty execution
+      final totalExecutionsDifficultyAmount = executionsAmounts[execution.markedDifficulty] ?? 0;
+      executionsAmounts[execution.markedDifficulty] = totalExecutionsDifficultyAmount + 1;
+    });
+
+    final userMetadata = associatedData[1] as MemoExecutionsMetadata;
+    final associatedCollection = associatedData[2] as Collection;
+    final userUpdatedExecutions = userMetadata.executionsAmounts;
+    final collectionUpdatedExecutions = associatedCollection.executionsAmounts;
+
+    // Uses the `executionsAmounts` to get the updated `executionsAmount` for the user and the associated collection
+    executionsAmounts.forEach((key, value) {
+      final userCurrentValue = userUpdatedExecutions[key] ?? 0;
+      userUpdatedExecutions[key] = userCurrentValue + value;
+
+      final collectionCurrentValue = collectionUpdatedExecutions[key] ?? 0;
+      collectionUpdatedExecutions[key] = collectionCurrentValue + value;
+    });
+
+    // Dispatch all updates simultaneously
```

Comment:
What happens if one fail? Shouldn't it be made using a transaction?

Reply:
This is a problem that I will have addressed in a future PR, but it's a valid proposition